### PR TITLE
[IMP][12.0] fieldservice_account

### PR DESCRIPTION
**Pull request by murtuzasaleh**
#391 

Some organizations allows their FSM persons to have access to the customer invoices.

### CODE REVIEW

**Comment by murtuzasaleh** [comment]
@max3903 Can you please review this PR.

**Review by osi-scampbell — APPROVED** [review]

**Comment by max3903** [comment]
@murtuzasaleh Let's got with a smart button instead of a tab and you are missing the security file to allow read access to invoices for FSM User.

File: fieldservice_account/views/fsm_order.xml [hunk 1]
```diff
@@ -7,6 +7,29 @@
...
+                            <field name="company_id" groups="base.group_multi_company" options="{'no_create': True}"/>
+                            <field name="user_id"/>
+                            <field name="date_due"/>
+                            <field name="origin"/>
+                            <field name="amount_untaxed_invoice_signed" string="Tax Excluded" sum="Total"/>
+                            <field name="amount_tax_signed" string="Tax" sum="Total"/>
+                            <field name="amount_total_signed" string="Total" sum="Total"/>
+                            <field name="residual_signed" string="Amount Due" sum="Amount Due"/>
+                            <field name="currency_id" invisible="1"/>
+                            <field name="state"/>
```
Comment (dreispt) [inline fieldservice_account/views/fsm_order.xml]: This is for information of the FSM Worker, so maybe we don't need all these fields here.
Implementation is correct, but we might want to refine the UX design in the future.

**Review by dreispt — APPROVED** [review]

**Comment by murtuzasaleh** [comment]
> @murtuzasaleh Let's got with a smart button instead of a tab and you are missing the security file to allow read access to invoices for FSM User.

The smart button will go again the requirement in #381 for the Order to have "Register Payment" buttons for the related Invoices. Should I then remove that code from the fieldservice_account_payment module?

File: fieldservice_account/views/fsm_order.xml (hunk 1 above)
Comment (dreispt) [reply fieldservice_account/views/fsm_order.xml]: For clarity, you can  leave it as it is, we can adjust later if needed.

File: fieldservice_account/models/account_invoice.py [hunk 2]
```diff
@@ -16,6 +16,7 @@ class AccountInvoice(models.Model):
     fsm_order_count = fields.Integer(
         string='FSM Order Count',
         compute='_compute_fsm_order_ids', readonly=True)
+    fsm_order_id = fields.Many2one('fsm.order', 'FSM Order')
```
Comment (max3903) [inline fieldservice_account/models/account_invoice.py]:
```suggestion
```
We already have `fsm_order_ids` above.

**Comment by max3903** [comment]
@wolfhall @hparfr @brian10048 @dreispt I am starting to get confused with the relationship between FSM orders and invoices. Couple commits ago we removed fsm_order_id from invoice and now we are adding it again.

Scenario 1:
We confirm a SO and it generates a delivery order, FSM orders and invoices. The FSM worker can get the payment of the invoice from the FSM order.

--> 1 invoice can be linked to multiple FSM orders.

Scenario 2:
When the FSM order is completed, the user can invoice the timesheets and contractor costs.

--> 1 FSM order can generate one invoice.

If we mix those 2 scenario, we can have 2 invoices for one FSM order (one for the inventory item being delivered, one for the contractor cost/timesheet of the installation).

What do you think of having a many2many relation, instead of 2 one2many ones?

**Comment by brian10048** [comment]
If there is one thing for certain we all have unique workflows for invoicing our field services.  If a many2many allows more flexibility for all scenarios, then that sounds good to me.

File: fieldservice_account/models/account_invoice.py (hunk 2 above)
Comment (max3903) [inline fieldservice_account/models/account_invoice.py]: @murtuzasaleh Remove this field and change fsm_order_ids above to a many2many.

File: fieldservice_account/models/fsm_order.py [hunk 3]
```diff
@@ -17,3 +17,7 @@ class FSMOrder(models.Model):
         related="invoice_line_id.invoice_id",
         readonly=True,
     )
+    invoice_ids = fields.One2many('account.invoice',
+                                  'fsm_order_id',
+                                  'Invoices'
+                                  )
```
Comment (max3903) [inline fieldservice_account/models/fsm_order.py]: @murtuzasaleh many2many as well

File: fieldservice_account/views/account_invoice.xml [hunk 4]
```diff
@@ -14,6 +14,9 @@
                     <field name="fsm_order_count" widget="statinfo" string="FSM Orders"/>
                 </button>
             </div>
+            <field name="reference" position="after">
+                <field name="fsm_order_id"/>
+            </field>
```
Comment (max3903) [inline fieldservice_account/views/account_invoice.xml]: @murtuzasaleh You can remove this. We will use the smart button.

File: fieldservice_account/views/fsm_order.xml [hunk 5]
```diff
@@ -7,6 +7,29 @@
...
+                        </tree>
+                    </field>
+                </page>
+            </notebook>
```
Comment (max3903) [inline fieldservice_account/views/fsm_order.xml]: @murtuzasaleh Let's use a smart button instead. Show it if there is at least one invoice.

**Comment by max3903** [comment]
@murtuzasaleh Please review and merge https://github.com/murtuzasaleh/field-service/pull/4

**Review by max3903 — APPROVED** [review]

**Comment by max3903** [comment]
@murtuzasaleh Can you provide the migration script please?

File: fieldservice_account/migrations/12.0.3.0.0/post-migration.py [hunk 6]
```diff
@@ -0,0 +1,13 @@
+# Copyright (C) 2019, Open Source Integrators
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+
+def migrate(env, version):
+    if not version:
+        return
+
+    env.execute("""INSERT INTO fsm_order_account_invoice_rel
+                SET (fsm_order_id, account_id) VALUES (
```
Comment (max3903) [inline fieldservice_account/migrations/12.0.3.0.0/post-migration.py]:
```suggestion
                SET (fsm_order_id, invoice_id) VALUES (
```

**Review by brian10048 — APPROVED** [review]

**Pull request merged**